proposals/records-wip.md

```diff
@@ -76,3 +76,45 @@ each primary constructor parameter to the corresponding property of the other ty
 ```C#
 override Equals(object o) => Equals(o as T);
 ```
+
+## `with` expression
+
+A `with` expression is a new expression using the following syntax.
+
+```antlr
+with_expression
+    : switch_expression
+    | switch_expression 'with' anonymous_object_initializer
+```
+
+A `with` expression allows for "non-destructive mutation", designed to
+produce a copy of the receiver expression with modifications to properties
+listed in the `anonymous_object_initializer`.
+
+A valid `with` expression has a receiver with a non-target-typed, non-null type. The receiver type must contain an instance method called `With` with
+the appropriate parameters and return type. It is an error if there are multiple non-override `With` methods. If there are multiple `With` overrides,
+there must be a non-override `With` method, which is the target method. Otherwise, there must be exactly one `With` method.
+
+On the right hand side of the `with` expression is an `anonymous_object_initializer` with a
+sequence of assignments with a field or property of the receiver on the left hand side of the
+assignment, and an arbitrary expression on the right hand side.
+
+Given a target `With` method, the return type must be the type of the receiver expression type, or a base type thereof. For each parameter of
+the `With` method, there must be a corresponding field or property on the
+receiver type with the same name and the same type.
```

Comment:
>same name and the same type [](start = 23, length = 27)

I'm confused as to why we're going with name matching. Mads was strongly opposed to that.

Reply:
Talked to him about this, he said it was fine for the prototype. I suspect this will not be the final design.